File: mobile/android/app/src/main/java/com/multimodaltrading/android/viewmodels/AuthViewModel.kt
package com.multimodaltrading.android.viewmodels

import androidx.lifecycle.ViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class AuthViewModel : ViewModel() {
    // State flows
    private val _isAuthenticated = MutableStateFlow(false)
    val isAuthenticated: StateFlow<Boolean> = _isAuthenticated.asStateFlow()
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    
    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()
    
    private val _user = MutableStateFlow<com.multimodaltrading.android.models.User?>(null)
    val user: StateFlow<com.multimodaltrading.android.models.User?> = _user.asStateFlow()
    
    // Check authentication status
    fun checkAuth() {
        // In a real implementation, this would check for stored credentials
        // For demo purposes, we'll just return false
        _isAuthenticated.value = false
    }
    
    // Login with email and password
    fun login(email: String, password: String) {
        _isLoading.value = true
        _error.value = null
        
        // In a real implementation, this would call the API
        // For demo purposes, we'll simulate a successful login
        
        // Simulate network delay
        android.os.Handler(android.os.Looper.getMainLooper()).postDelayed({
            _isLoading.value = false
            
            if (email.isNotEmpty() && password.isNotEmpty()) {
                _isAuthenticated.value = true
                _user.value = com.multimodaltrading.android.models.User(
                    id = "1",
                    email = email,
                    name = "Demo User"
                )
            } else {
                _error.value = "Invalid email or password"
            }
        }, 1000)
    }
    
    // Connect wallet
    fun connectWallet(walletType: String) {
        _isLoading.value = true
        _error.value = null
        
        // In a real implementation, this would connect to the wallet
        // For demo purposes, we'll simulate a successful connection
        
        // Simulate network delay
        android.os.Handler(android.os.Looper.getMainLooper()).postDelayed({
            _isLoading.value = false
            _isAuthenticated.value = true
            _user.value = com.multimodaltrading.android.models.User(
                id = "wallet_${walletType}_1",
                email = null,
                name = "${walletType.capitalize()} User"
            )
        }, 1000)
    }
    
    // Demo login
    fun demoLogin() {
        _isLoading.value = true
        _error.value = null
        
        // Simulate network delay
        android.os.Handler(android.os.Looper.getMainLooper()).postDelayed({
            _isLoading.value = false
            _isAuthenticated.value = true
            _user.value = com.multimodaltrading.android.models.User(
                id = "demo_1",
                email = "demo@example.com",
                name = "Demo User"
            )
        }, 500)
    }
    
    // Connect (general method)
    fun connect() {
        // In a real implementation, this would show wallet options
        // For demo purposes, we'll just use the demo login
        demoLogin()
    }
    
    // Disconnect
    fun disconnect() {
        _isAuthenticated.value = false
        _user.value = null
    }
    
    // Extension function to capitalize first letter
    private fun String.capitalize(): String {
        return this.replaceFirstChar { if (it.isLowerCase()) it.titlecase() else it.toString() }
    }
}

File: mobile/ios/MultimodalTrading/ViewModels/AuthViewModel.swift
import Foundation
import Combine

class AuthViewModel: ObservableObject {
    @Published var isAuthenticated: Bool = false
    @Published var isLoading: Bool = false
    @Published var error: String? = nil
    @Published var user: User? = nil
    
    private var cancellables = Set<AnyCancellable>()
    
    // Login with email and password
    func login(email: String, password: String) {
        isLoading = true
        error = nil
        
        APIService.shared.login(email: email, password: password)
            .sink(receiveCompletion: { [weak self] completion in
                self?.isLoading = false
                
                if case .failure(let error) = completion {
                    self?.error = error.localizedDescription
                }
            }, receiveValue: { [weak self] response in
                self?.handleSuccessfulAuth(response)
            })
            .store(in: &cancellables)
    }
    
    // Connect wallet
    func connectWallet(type: String) {
        isLoading = true
        error = nil
        
        // In a real implementation, this would use a wallet SDK
        // For demo purposes, we'll simulate a successful connection
        
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) { [weak self] in
            // Generate a random wallet address
            let walletAddress = "sol" + String(UUID().uuidString.prefix(8))
            let signature = UUID().uuidString
            
            // Call API to authenticate with wallet
            APIService.shared.connectWallet(address: walletAddress, signature: signature)
                .sink(receiveCompletion: { [weak self] completion in
                    self?.isLoading = false
                    
                    if case .failure(let error) = completion {
                        self?.error = error.localizedDescription
                    }
                }, receiveValue: { [weak self] response in
                    self?.handleSuccessfulAuth(response)
                })
                .store(in: &(self?.cancellables ?? []))
        }
    }
    
    // Demo login (no API call)
    func demoLogin() {
        isLoading = true
        error = nil
        
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) { [weak self] in
            // Create demo user
            let user = User(id: "demo-user", email: "demo@example.com", walletAddress: "sol1234demo")
            
            // Save token
            UserDefaults.standard.set("demo-token", forKey: "auth_token")
            
            // Update state
            self?.isAuthenticated = true
            self?.user = user
            self?.isLoading = false
        }
    }
    
    // Connect (generic method)
    func connect() {
        // Show wallet options or login form
        // For demo purposes, we'll just use demo login
        demoLogin()
    }
    
    // Disconnect
    func disconnect() {
        // Clear token
        UserDefaults.standard.removeObject(forKey: "auth_token")
        
        // Update state
        isAuthenticated = false
        user = nil
    }
    
    // Handle successful authentication
    private func handleSuccessfulAuth(_ response: AuthResponse) {
        // Save token
        UserDefaults.standard.set(response.token, forKey: "auth_token")
        
        // Update state
        isAuthenticated = true
        user = response.user
    }
    
    // Check if token exists on app launch
    func checkAuth() {
        if let token = UserDefaults.standard.string(forKey: "auth_token") {
            // Token exists, validate it
            isAuthenticated = true
            
            // In a real implementation, this would validate the token with the API
            // For demo purposes, we'll just create a demo user
            user = User(id: "demo-user", email: "demo@example.com", walletAddress: "sol1234demo")
        }
    }
}
